File: src/explainable_engine/models/response.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class BreakdownItem(BaseModel):
    node_id: str
    label: str
    value: float
    weight: float
    absolute_contribution: float
    percentage: float
    confidence: float
    children: list[BreakdownItem] | None = None


class DriverItem(BaseModel):
    name: str
    impact: float
    rank: int


class GraphNodeResponse(BaseModel):
    id: str
    label: str
    value: float
    confidence: float
    node_type: str


class GraphEdgeResponse(BaseModel):
    source: str
    target: str
    weight: float
    transformation_type: str


class GraphResponse(BaseModel):
    nodes: list[GraphNodeResponse]
    edges: list[GraphEdgeResponse]


class DependencyNode(BaseModel):
    node_id: str
    label: str
    depth: int
    relation: str | None = None
    children: list[DependencyNode] = Field(default_factory=list)


class DependencyTree(BaseModel):
    root: DependencyNode
    depth: int
    total_nodes: int


class ConfidenceDetail(BaseModel):
    overall: float
    per_node: dict[str, float]


class ExplainMetadata(BaseModel):
    version: str
    created_at: datetime
    deterministic_hash: str
    computation_type: str = "additive"


class ExplainResponse(BaseModel):
    id: str
    target: str
    final_value: float
    confidence: float
    breakdown: list[BreakdownItem]
    top_drivers: list[DriverItem]
    missing_impact: float = 0.0
    graph: GraphResponse | None = None
    dependency_tree: DependencyTree | None = None
    confidence_detail: ConfidenceDetail | None = None
    metadata: ExplainMetadata

File: internal/models/response.go
package models

import "time"

// BreakdownItem represents one component's contribution to the result.
type BreakdownItem struct {
	NodeID               string          `json:"node_id"`
	Label                string          `json:"label"`
	Value                float64         `json:"value"`
	Weight               float64         `json:"weight"`
	AbsoluteContribution float64         `json:"absolute_contribution"`
	Percentage           float64         `json:"percentage"`
	Confidence           float64         `json:"confidence"`
	Children             []BreakdownItem `json:"children,omitempty"`
}

// DriverItem represents a top driver of the result.
type DriverItem struct {
	Name   string  `json:"name"`
	Impact float64 `json:"impact"`
	Rank   int     `json:"rank"`
}

// GraphNodeResponse is the API representation of a graph node.
type GraphNodeResponse struct {
	ID         string  `json:"id"`
	Label      string  `json:"label"`
	Value      float64 `json:"value"`
	Confidence float64 `json:"confidence"`
	NodeType   string  `json:"node_type"`
}

// GraphEdgeResponse is the API representation of a graph edge.
type GraphEdgeResponse struct {
	Source             string  `json:"source"`
	Target             string  `json:"target"`
	Weight             float64 `json:"weight"`
	TransformationType string  `json:"transformation_type"`
}

// GraphResponse contains the full graph representation.
type GraphResponse struct {
	Nodes []GraphNodeResponse `json:"nodes"`
	Edges []GraphEdgeResponse `json:"edges"`
}

// DependencyNode represents a node in the dependency tree.
type DependencyNode struct {
	NodeID   string           `json:"node_id"`
	Label    string           `json:"label"`
	Depth    int              `json:"depth"`
	Relation string           `json:"relation,omitempty"`
	Children []DependencyNode `json:"children,omitempty"`
}

// DependencyTree represents the full dependency tree.
type DependencyTree struct {
	Root       DependencyNode `json:"root"`
	Depth      int            `json:"depth"`
	TotalNodes int            `json:"total_nodes"`
}

// ConfidenceDetail contains per-node confidence information.
type ConfidenceDetail struct {
	Overall float64            `json:"overall"`
	PerNode map[string]float64 `json:"per_node"`
}

// ExplainMetadata contains metadata about the explanation.
type ExplainMetadata struct {
	Version           string    `json:"version"`
	CreatedAt         time.Time `json:"created_at"`
	DeterministicHash string    `json:"deterministic_hash"`
	ComputationType   string    `json:"computation_type"`
}

// ExplainResponse is the full explanation output.
type ExplainResponse struct {
	ID               string            `json:"id"`
	Target           string            `json:"target"`
	FinalValue       float64           `json:"final_value"`
	Confidence       float64           `json:"confidence"`
	Breakdown        []BreakdownItem   `json:"breakdown"`
	TopDrivers       []DriverItem      `json:"top_drivers"`
	MissingImpact    float64           `json:"missing_impact"`
	Graph            *GraphResponse    `json:"graph,omitempty"`
	DependencyTree   *DependencyTree   `json:"dependency_tree,omitempty"`
	ConfidenceDetail *ConfidenceDetail `json:"confidence_detail,omitempty"`
	Metadata         ExplainMetadata   `json:"metadata"`
	OriginalRequest  *ExplainRequest   `json:"original_request,omitempty"`
}
